### PR TITLE
Socket CAN sample and driver

### PR DESCRIPTION
**Pull request by jukkar**
This contains the changes to stm32 CAN driver to support the socket CAN. It contains also the socket CAN sample application. The code was originally in #12779 but is split as this might get longer time to merge.

### CODE REVIEW

**Review by pfalcon — COMMENTED** [review]
Thanks for splitting large PR.

Following https://github.com/zephyrproject-rtos/zephyr/pull/12779#discussion_r251734629, can the sample be made to run on both Zephyr and Linux (with the same code, perhaps different header inclusions), and instructions added how to run it on Linux using dummy CAN driver they have?

**Comment by jukkar** [comment]
Fixed merge conflicts and added filter setting support to sample application.

File: drivers/can/socket_can_generic.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+CAN_DEFINE_MSGQ(socket_can_msgq, 5);
+K_THREAD_STACK_DEFINE(rx_thread_stack, RX_THREAD_STACK_SIZE);
+
+struct socket_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	struct k_msgq *msgq;
+	k_tid_t rx_tid;
+	struct k_thread rx_thread_data;
+	struct can_filter filter;
```
Comment (alexanderwachter) [inline drivers/can/socket_can_generic.h]: Why do you need a filter in socket_can_context?

File: drivers/can/socket_can_generic.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	if (optlen != sizeof(struct can_filter)) {
+		errno = -EINVAL;
+		return -1;
+	}
+
+	memcpy(&socket_context->filter, optval, sizeof(struct can_filter));
+
+	can_attach_msgq(socket_context->can_dev,
+			socket_context->msgq,
+			&socket_context->filter);
```
Comment (alexanderwachter) [inline drivers/can/socket_can_generic.h]: you can pass optval direct here.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/can/socket_can_generic.h (hunk 1 above)
Comment (jukkar) [reply drivers/can/socket_can_generic.h]: I was not sure what the CAN driver will do with the filter. If the driver just copies the filter or does not refer to the original buffer, then sure the extra variable is not needed. I will remove it.

File: drivers/can/socket_can_generic.h (hunk 2 above)
Comment (jukkar) [reply drivers/can/socket_can_generic.h]: sure

File: drivers/can/socket_can_generic.h (hunk 1 above)
Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: The filter is just a generic struct that represents filters. The driver takes this information and sets the registers accordingly. Its save to discard after the call.

File: drivers/can/socket_can_generic.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+static struct canbus_api socket_can_api = {
+	.iface_api.init = socket_can_iface_init,
+	.send = socket_can_send,
+	.setsockopt = socket_can_setsockopt,
+};
+
+static struct socket_can_context socket_can_context_1;
+
+static inline void rx_thread(void *ctx, void *unused1, void *unused2)
+{
```
Comment (alexanderwachter) [inline drivers/can/socket_can_generic.h]: Do we really need a thread for receiving? The data processing is anyway offloaded to a work queue by net_recv_data() -> net_queue_rx(). When having a look at the Ethernet drivers, there are some having a thread (stm32) and some without (dw). I would prefer not to have the overhead of yet another thread if possible.

Comment (jukkar) [reply drivers/can/socket_can_generic.h]: Difficult to know atm about the time spent on ISR. My stm32f072b_disco just gives error after few packets. Anyway, what you are suggesting sounds better than having a thread from memory consumption point of view. And as having K_NO_WAIT there is ok as the CAN traffic is not really meant to be reliable. Would you be ok if we try to go with the thread solution first, and change this to using isr a bit later. I am hoping we could merge this by end of week so that we could get this in 1.14.

Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: Sure. This was just a suggestion. I was just wondering about the inconsistency in the Ethernet drivers some time ago.
In my opinion CAN traffic should be reliable. The datalink layer itself is very reliable. Where did you get the errors? Did you had an overload in the controller fifo?

Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: Regarding FIFO overrun: Maybe I should check for overruns in the device drivers.

File: subsys/net/lib/sockets/sockets_packet.c [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	/* Register the callback before sending in order to receive the response
+	 * from the peer.
+	 */
+
+	status = net_context_recv(ctx, zpacket_received_cb, K_NO_WAIT,
```
Comment (alexanderwachter) [inline subsys/net/lib/sockets/sockets_packet.c]: Why do we need to register the zpacket_received_cb here?

Comment (jukkar) [reply subsys/net/lib/sockets/sockets_packet.c]: That is part of the packet socket patch that I had to build on top of. So that is discussed on #12564. I copied the same code to sockets_can.c but it might be that we do not need this call after all. That needs more investigation.

**Comment by jukkar** [comment]
Updated according to comments.

**Comment by jukkar** [comment]
Tested the sample and it is able to send and receive data in loopback mode.

File: samples/net/sockets/packet/README.rst [hunk 5]
```diff
@@ -0,0 +1,28 @@
+.. _packet-socket-sample:
+
+Raw socket sample
+#################
+
+Overview
+********
+
+This sample is a simple packet socket application showing usage
+of packet socket over ethernet. Sample prints every packet which
```
Comment (dbkinder) [inline samples/net/sockets/packet/README.rst]: ``Ethernet`` is a trademarked name, so needs to be upper case.

Here's a rewrite of this paragraph with some wording edits:
```
This sample is a simple packet socket application showing usage
of packet sockets over Ethernet. The sample prints every packet
received, and sends a dummy packet every 5 seconds.
The Zephyr network subsystem does not touch any of the headers
(L2, L3, etc.). 
```

File: samples/net/sockets/packet/README.rst [hunk 6]
```diff
@@ -0,0 +1,28 @@
+.. _packet-socket-sample:
+
+Raw socket sample
```
Comment (dbkinder) [inline samples/net/sockets/packet/README.rst]: Should this doc title be "Packet socket sample" (to match the label above and the body text below)?

**Review by dbkinder — DISMISSED** [review]

File: samples/net/sockets/packet/README.rst (hunk 5 above)
Comment (rveerama1) [reply samples/net/sockets/packet/README.rst]: Comments are addressed in original PR https://github.com/zephyrproject-rtos/zephyr/pull/12564

File: samples/net/sockets/packet/README.rst (hunk 6 above)
Comment (rveerama1) [reply samples/net/sockets/packet/README.rst]: Comments are addressed in original PR https://github.com/zephyrproject-rtos/zephyr/pull/12564, thanks.

**Comment by jukkar** [comment]
recheck

**Comment by jukkar** [comment]
@alexanderwachter could you recheck the driver changes? There are known fixes like the dispatcher support coming a bit later.

**Review by alexanderwachter — APPROVED** [review]

**Review by pfalcon — CHANGES_REQUESTED** [review]
Review queue.

**Comment by jukkar** [comment]
updated the sample.yaml

**Comment by jukkar** [comment]
ping @pfalcon

**Comment by pfalcon** [comment]
First reviewing https://github.com/zephyrproject-rtos/zephyr/pull/12779 which was sneak-merged.

**Comment by pfalcon** [comment]
Please fix the issue from the previous patchset in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/12779/commits/792ceb508892a0777483cd25de2a6d5df104a8a1#r255031943

> This is the same problem as with initial packet socket patch. Please dispatch the entire address family to your implementation, and deal with different socket types within it there. Please fix this in #12808

File: drivers/can/socket_can_generic.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+#define BUF_ALLOC_TIMEOUT K_MSEC(50)
+
+/* TODO: make msgq size configurable */
+CAN_DEFINE_MSGQ(socket_can_msgq, 5);
+K_THREAD_STACK_DEFINE(rx_thread_stack, RX_THREAD_STACK_SIZE);
+
+struct socket_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	struct k_msgq *msgq;
```
Comment (pfalcon) [inline drivers/can/socket_can_generic.h]: So, here's exactly what I mean - for CAN, you have your own queues and infrastructure anyway. Sockets could work with this infrastructure directly. Instead, `net/ip/` code was slashed and twisted to useless plug all this unrelated CAN stuff into it.

**Review by pfalcon — CHANGES_REQUESTED** [review]
1. Some issues from the previous patchset, hastily merged, need to be fixed first. (See comments.)
2. There should be a sample which runs both on Linux and Zephyr.

File: drivers/can/socket_can_generic.h (hunk 7 above)
Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: It would actually be great when all raw sockets can bypass the IP stack somehow. This would reduce the memory footprint a lot.

Comment (pfalcon) [reply drivers/can/socket_can_generic.h]: > when all raw sockets can bypass the IP stack somehow.

Well, "all raw sockets" can't bypass IP stack, because there're IP raw sockets. But sockets of address families not related to IP (e.g., AF_PACKET, AF_CAN) definitely can bypass IP stack, and it would make the IP stack much more manageable (given that we discover bug after bug in it).

File: drivers/can/socket_can_generic.h (hunk 3 above)
Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: "Transmit buffer full. Wait with timeout (100ms)" is just a DBG message. This means all of the three transmit mailboxes are full. A tx interrupt will then trigger the tx_int_sem and sending goes on. This is nothing to worry about.

File: drivers/can/socket_can_generic.h (hunk 7 above)
Comment (alexanderwachter) [reply drivers/can/socket_can_generic.h]: Regarding the msgq: As already discussed above, this will be removed and the isr will be used instead. For now we have two threads that pass the same massage.

**Comment by jukkar** [comment]
New version which adds support for nucleo_l432kc in the sample.

**Comment by jukkar** [comment]
> * Some issues from the previous patchset, hastily merged, need to be fixed first. (See comments.)

These two patchsets are not too much related, so future changes do not affect this one.

> 
> * There should be a sample which runs both on Linux and Zephyr.

Currently this is not done, reason being:
* I do not have setup for testing CAN on linux
* Low level CAN headers etc. are not compatible with linux and zephyr, so would require #ifdefs etc. Perhaps it makes sense to do it or not, difficult to say yet.

**Comment by pfalcon** [comment]
> Currently this is not done, reason being:
>    * I do not have setup for testing CAN on linux

So, I mean, a sample using dummy CAN driver, see "6.4 The virtual CAN driver (vcan)" https://www.kernel.org/doc/Documentation/networking/can.txt

>     * Low level CAN headers etc. are not compatible with linux and zephyr, so would require #ifdefs etc. Perhaps it makes sense to do it or not, difficult to say yet.

So, having a sample running on Linux is exactly needed to catch things like that. Because we develop, develop standard APIs, and in the end turns out that it doesn't work as expected.

In that specific case, who's wrong? Even if it's Linux, I'd suggest to just use Linux's stuff, unless it's *too* bloated.

**Review by pfalcon — APPROVED** [review]
Recommended to explore a sample compatible with both Linux and Zephyr, using Linux "vcan" module as the next step.

File: drivers/can/socket_can_generic.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	int ret;
+
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+
+	while (1) {
+		k_msgq_get((struct k_msgq *)socket_context->msgq, &msg,
+			   K_FOREVER);
+
+		pkt = net_pkt_get_reserve_rx(BUF_ALLOC_TIMEOUT);
```
Comment (tbursztyka) [inline drivers/can/socket_can_generic.h]: Use the new API, see for instance drivers/ethernet/eth_dw.c 

Should be trivial to fix.

File: drivers/can/socket_can_generic.h [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		if (pkt == NULL) {
+			LOG_ERR("Cannot alloc buf");
+			continue;
+		}
+
+		ret = net_pkt_append_all(pkt, sizeof(msg), (const u8_t *)&msg,
```
Comment (tbursztyka) [inline drivers/can/socket_can_generic.h]: same here, net_pkt_write_new() will do the job.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Just missing to use the new net_pkt API. Rest is fine to me.

**Comment by alexanderwachter** [comment]
Regarding the different headers:
- Linux:
``` C
struct can_frame {
	canid_t can_id;  /* 32 bit CAN_ID + EFF/RTR/ERR flags */
	__u8    can_dlc; /* frame payload length in byte (0 .. CAN_MAX_DLEN) */
	__u8    __pad;   /* padding */
	__u8    __res0;  /* reserved / padding */
	__u8    __res1;  /* reserved / padding */
	__u8    data[CAN_MAX_DLEN] __attribute__((aligned(8)));
};
```
 - Zephyr:
``` C
struct can_msg {
	/** Indicates the identifier type (standard or extended)
	 * use can_ide enum for assignment
	 */
	u32_t id_type : 1;
	/** Set the message to a transmission request instead of data frame
	 * use can_rtr enum for assignment
	 */
	u32_t rtr     : 1;
	/** Message identifier*/
	union {
		u32_t std_id  : 11;
		u32_t ext_id  : 29;
	};
	/** The length of the message (max. 8) in byte */
	u8_t dlc;
	/** The message data*/
	union {
		u8_t data[8];
		u32_t data_32[2];
	};
} __packed;
```
I dind't had any capability in mind while writing the CAN API. If we want to be compatible we need to change that or translate the header in socket can

**Comment by pfalcon** [comment]
> If we want to be compatible we need to change that or translate the header in socket can

So, I'd recommend going that way (using just 1 struct everywhere, compatible with Linux). Just using compatible field names and types (then macros to access bitfields in them, apparently). As for res/pad, we don't need to have them explicitly, but of course we'll have them anyway due to alignment.

**Comment by alexanderwachter** [comment]
@jukkar What do you think? Can you implement this? I'm busy with ISO-TP already

File: drivers/can/socket_can_generic.h (hunk 8 above)
Comment (jukkar) [reply drivers/can/socket_can_generic.h]: done

File: drivers/can/socket_can_generic.h (hunk 9 above)
Comment (jukkar) [reply drivers/can/socket_can_generic.h]: done

**Comment by jukkar** [comment]
> What do you think? Can you implement this? I'm busy with ISO-TP already

Sure. This is not super critical atm, I can look it next week.

**Review by tbursztyka — APPROVED** [review]
